homeassistant/components/shelly/__init__.py

```diff
@@ -369,42 +407,128 @@ def mac(self) -> str:
     if device is not None:
         # If device is present, device wrapper is not setup yet
         device.shutdown()
         return True
 
-    platforms = SLEEPING_PLATFORMS
+    platforms = BLOCK_SLEEPING_PLATFORMS
 
     if not entry.data.get("sleep_period"):
         hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry.entry_id][REST] = None
-        platforms = PLATFORMS
+        platforms = BLOCK_PLATFORMS
 
     unload_ok = await hass.config_entries.async_unload_platforms(entry, platforms)
     if unload_ok:
-        hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry.entry_id][COAP].shutdown()
+        hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry.entry_id][BLOCK].shutdown()
         hass.data[DOMAIN][DATA_CONFIG_ENTRY].pop(entry.entry_id)
 
     return unload_ok
 
 
 def get_device_wrapper(
     hass: HomeAssistant, device_id: str
-) -> ShellyDeviceWrapper | None:
+) -> BlockDeviceWrapper | RpcDeviceWrapper | None:
     """Get a Shelly device wrapper for the given device id."""
     if not hass.data.get(DOMAIN):
         return None
 
     for config_entry in hass.data[DOMAIN][DATA_CONFIG_ENTRY]:
-        wrapper: ShellyDeviceWrapper | None = hass.data[DOMAIN][DATA_CONFIG_ENTRY][
+        block_wrapper: BlockDeviceWrapper | None = hass.data[DOMAIN][DATA_CONFIG_ENTRY][
+            config_entry
+        ].get(BLOCK)
+
+        if block_wrapper and block_wrapper.device_id == device_id:
+            return block_wrapper
+
+        rpc_wrapper: RpcDeviceWrapper | None = hass.data[DOMAIN][DATA_CONFIG_ENTRY][
             config_entry
-        ].get(COAP)
+        ].get(RPC)
 
-        if wrapper and wrapper.device_id == device_id:
-            return wrapper
+        if rpc_wrapper and rpc_wrapper.device_id == device_id:
+            return rpc_wrapper
 
     return None
+
+
+class RpcDeviceWrapper(update_coordinator.DataUpdateCoordinator):
+    """Wrapper for a Shelly RPC based device with Home Assistant specific functions."""
+
+    def __init__(
+        self, hass: HomeAssistant, entry: ConfigEntry, device: RpcDevice
+    ) -> None:
+        """Initialize the Shelly device wrapper."""
+        self.device_id: str | None = None
+
+        device_name = get_rpc_device_name(device) if device.initialized else entry.title
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=device_name,
+            update_interval=timedelta(seconds=RPC_RECONNECT_INTERVAL),
+        )
+        self.entry = entry
+        self.device = device
+
+        entry.async_on_unload(
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self._handle_ha_stop)
+        )
+
+    async def _async_update_data(self) -> None:
+        """Fetch data."""
+        if self.device.connected:
+            return
+
+        try:
+            _LOGGER.debug("Reconnecting to Shelly RPC Device - %s", self.name)
+            async with async_timeout.timeout(AIOSHELLY_DEVICE_TIMEOUT_SEC):
+                await self.device.initialize()
```

Comment:
Does this mean we can call initialize multiple times if we lose connection?

Reply:
Yes